### PR TITLE
Who has perm + notifications

### PR DESCRIPTION
**Pull request by a-belhadj**

### CODE REVIEW

File: profiles/models/request_notification.py [hunk 1]
```diff
@@ -40,8 +40,10 @@ def is_authorized(self, request):
 
     def when_render(self, request):
         from service_catalog.api.serializers import RequestSerializer
+        from service_catalog.api.serializers import InstanceSerializer
         context = {
-            "request": RequestSerializer(request).data
+            "request": RequestSerializer(request).data,
+            "instance": InstanceSerializer(request.instance).data
```
Comment (EliasBoulharts) [inline profiles/models/request_notification.py:46]: The instance serializer is in the request serializer, if I'm not wrong

File: service_catalog/api/views/approval_workflow_api_views.py [hunk 2]
```diff
@@ -49,7 +50,7 @@ def get_survey(self, request, pk=None):
         target_request = get_object_or_404(Request, id=pk)
         if target_request.approval_workflow_state.current_step is None:
             return Response({"error": "No pending step to approve"}, status=HTTP_404_NOT_FOUND)
-        if not request.user.has_perm(target_request.approval_workflow_state.current_step.approval_step.permission,
+        if not request.user.has_perm(target_request.approval_workflow_state.current_step.approval_step.permission.get_permission_str,
```
Comment (EliasBoulharts) [inline service_catalog/api/views/approval_workflow_api_views.py:53]: get_permission_str is a method but it should be a property IMO (permission_str without get)

File: service_catalog/api/views/approval_workflow_api_views.py [hunk 3]
```diff
@@ -62,14 +63,14 @@ def approve(self, request, pk=None):
         target_request = get_object_or_404(Request, id=pk)
         if target_request.approval_workflow_state.current_step is None:
             return Response({"error": "No pending step to approve"}, status=HTTP_404_NOT_FOUND)
-        if not request.user.has_perm(target_request.approval_workflow_state.current_step.approval_step.permission,
+        if not request.user.has_perm(target_request.approval_workflow_state.current_step.approval_step.permission.get_permission_str(),
```
Comment (EliasBoulharts) [inline service_catalog/api/views/approval_workflow_api_views.py:66]: Here it is called as method

File: service_catalog/api/views/approval_workflow_api_views.py [hunk 4]
```diff
@@ -78,14 +79,16 @@ def reject(self, request, pk=None):
...
```
Comment (EliasBoulharts) [inline service_catalog/api/views/approval_workflow_api_views.py:82]: Same as above

File: service_catalog/api/views/approval_workflow_api_views.py [hunk 5]
```diff
@@ -78,14 +79,16 @@ def reject(self, request, pk=None):
...
                                      target_request):
             raise PermissionDenied
-        target_request.approval_workflow_state.reject_current_step(request.user)
+        # reject the current step if exist
+        if target_request.approval_workflow_state is not None:
+            target_request.approval_workflow_state.reject_current_step(request.user)
         # reject the request
         target_request.reject(request.user)
         target_request.save()
-        # send_mail_request_update(target_request, user_applied_state=request.user, message=message)
```
Comment (EliasBoulharts) [inline service_catalog/api/views/approval_workflow_api_views.py:88]: message are no more displayed in notifications?

File: tests/test_service_catalog/test_api/test_approval/test_approval_workflow_api_views.py [hunk 6]
```diff
@@ -66,3 +69,104 @@ def test_set_position_missing_position(self):
...
+                                                     requester=self.standard_user,
+                                                     quota_scope=self.test_quota_scope)
+        data = {
+            'text_variable': 'my_var',
+            'multiplechoice_variable': 'choice1', 'multiselect_var': 'multiselect_1',
+            'textarea_var': '2',
+            'password_var': 'pass',
+            'integer_var': '1',
+            'float_var': '0.6'
+        }
```
Comment (EliasBoulharts) [inline tests/test_service_catalog/test_api/test_approval/test_approval_workflow_api_views.py:104]: variable used 1 time can be removed

File: tests/test_service_catalog/test_api/test_approval/test_approval_workflow_api_views.py [hunk 7]
```diff
@@ -66,3 +69,104 @@ def test_set_position_missing_position(self):
...
+        self.test_request = Request.objects.create(fill_in_survey=data,
+                                                   instance=self.test_instance,
+                                                   operation=self.create_operation_test,
+                                                   user=self.standard_user)
+
+    def test_superuser_can_approve(self):
+
```
Comment (EliasBoulharts) [inline tests/test_service_catalog/test_api/test_approval/test_approval_workflow_api_views.py:111]: space

File: tests/test_service_catalog/test_api/test_approval/test_approval_workflow_api_views.py [hunk 8]
```diff
@@ -66,3 +69,104 @@ def test_set_position_missing_position(self):
...
+    def test_superuser_can_reject(self):
+
+
+        self.client.force_login(self.superuser)
+        url_reject = reverse('api_request_approval_workflow_state_reject', kwargs={'pk': self.test_request.pk})
+        response = self.client.post(url_reject, format='json')
+        self.test_request.refresh_from_db()
+        self.assertTrue(self.test_request.state == RequestState.REJECTED)
+        self.assertEqual(response.status_code, 200)
+    def test_approver_can_approve(self):
```
Comment (EliasBoulharts) [inline tests/test_service_catalog/test_api/test_approval/test_approval_workflow_api_views.py:128]: no space

File: tests/test_service_catalog/test_api/test_approval/test_approval_workflow_api_views.py [hunk 9]
```diff
@@ -66,3 +69,104 @@ def test_set_position_missing_position(self):
...
+
```
Comment (EliasBoulharts) [inline tests/test_service_catalog/test_api/test_approval/test_approval_workflow_api_views.py:129]: space

File: tests/test_service_catalog/test_api/test_approval/test_approval_workflow_api_views.py [hunk 10]
```diff
@@ -66,3 +69,104 @@ def test_set_position_missing_position(self):
...
+        self.test_quota_scope.add_user_in_role(approver, role_approver)
+        self.client.force_login(approver)
+        url_approve = reverse('api_request_approval_workflow_state_approve', kwargs={'pk': self.test_request.pk})
+        response = self.client.post(url_approve, format='json')
+        self.test_request.refresh_from_db()
+        self.assertTrue(self.test_request.state == RequestState.ACCEPTED)
+        self.assertEqual(response.status_code, 200)
+
+    def test_approver_can_reject(self):
+
```
Comment (EliasBoulharts) [inline tests/test_service_catalog/test_api/test_approval/test_approval_workflow_api_views.py:142]: space

File: tests/test_service_catalog/test_api/test_approval/test_approval_workflow_api_views.py [hunk 11]
```diff
@@ -66,3 +69,104 @@ def test_set_position_missing_position(self):
...
+        self.client.force_login(approver)
+        url_reject = reverse('api_request_approval_workflow_state_reject', kwargs={'pk': self.test_request.pk})
+        response = self.client.post(url_reject, format='json')
+        self.test_request.refresh_from_db()
+        self.assertTrue(self.test_request.state == RequestState.REJECTED)
+        self.assertEqual(response.status_code, 200)
+
+
+    def test_standard_user_cannot_approve(self):
+
```
Comment (EliasBoulharts) [inline tests/test_service_catalog/test_api/test_approval/test_approval_workflow_api_views.py:157]: space

File: tests/test_service_catalog/test_api/test_approval/test_approval_workflow_api_views.py [hunk 12]
```diff
@@ -66,3 +69,104 @@ def test_set_position_missing_position(self):
...
+        self.client.force_login(self.standard_user)
+        url_approve = reverse('api_request_approval_workflow_state_approve', kwargs={'pk': self.test_request.pk})
+        response = self.client.post(url_approve, format='json')
+        self.test_request.refresh_from_db()
+        self.assertEqual(response.status_code, 403)
+        self.assertTrue(self.test_request.state == RequestState.SUBMITTED)
+
+    def test_standard_user_cannot_reject(self):
+
```
Comment (EliasBoulharts) [inline tests/test_service_catalog/test_api/test_approval/test_approval_workflow_api_views.py:166]: space

File: tests/test_mail_utils.py [hunk 13]
```diff
@@ -114,29 +142,74 @@ def test_get_subject(self):
...
+                                                   user=self.standard_user)
+
+    def test_get_admin_emails_with_request(self):
+        approver = User.objects.create(username='approver', email="approver@local.com")
+        role_approver = Role.objects.create(name="Approver")
+        role_approver.permissions.add(self.approval_step1.permission)
+
+        self.test_quota_scope.add_user_in_role(approver,role_approver)
+        self.assertCountEqual([self.superuser.email, self.superuser_2.email,approver.email],
+                              _get_receivers_for_request(self.test_request))
```
Comment (EliasBoulharts) [inline tests/test_mail_utils.py:215]: put in another file?

File: tests/test_mail_utils.py [hunk 14]
```diff
@@ -114,29 +142,74 @@ def test_get_subject(self):
...
+                                                     requester=self.standard_user,
+                                                     quota_scope=self.test_quota_scope)
+        data = {
+            'text_variable': 'my_var',
+            'multiplechoice_variable': 'choice1', 'multiselect_var': 'multiselect_1',
+            'textarea_var': '2',
+            'password_var': 'pass',
+            'integer_var': '1',
+            'float_var': '0.6'
+        }
```
Comment (EliasBoulharts) [inline tests/test_mail_utils.py:202]: used 1 time can be removed

File: service_catalog/mail_utils.py [hunk 15]
```diff
@@ -37,36 +25,61 @@ def _get_headers(subject):
...
-                    email_admins.append(admin.email)
-            elif isinstance(object_to_filter, Request) and admin.profile.request_notification_enabled:
-                if admin.profile.is_notification_authorized_for_request(object_to_filter):
-                    email_admins.append(admin.email)
-    return email_admins
+def _apply_when_filter_instance(user_list, squest_object):
+    users = list()
+    for user in user_list:
+        if user.email:
+            if user.profile.instance_notification_enabled:
```
Comment (EliasBoulharts) [inline service_catalog/mail_utils.py:32]: So, an User queryset can be filtered easly

File: service_catalog/mail_utils.py [hunk 16]
```diff
@@ -37,36 +25,61 @@ def _get_headers(subject):
...
+                if user.profile.is_notification_authorized_for_instance(squest_object):
+                    users.append(user)
+    return users
+
+
+def _apply_when_filter_request(user_list, squest_object):
+    users = list()
+    for user in user_list:
+        if user.email:
+            if user.profile.request_notification_enabled:
```
Comment (EliasBoulharts) [inline service_catalog/mail_utils.py:42]: same as above

File: service_catalog/mail_utils.py [hunk 17]
```diff
@@ -37,36 +25,61 @@ def _get_headers(subject):
...
-                if admin.profile.is_notification_authorized_for_request(object_to_filter):
-                    email_admins.append(admin.email)
-    return email_admins
+def _apply_when_filter_instance(user_list, squest_object):
+    users = list()
+    for user in user_list:
+        if user.email:
+            if user.profile.instance_notification_enabled:
+                if user.profile.is_notification_authorized_for_instance(squest_object):
+                    users.append(user)
```
Comment (EliasBoulharts) [inline service_catalog/mail_utils.py:34]: why do you not build the list of email here?

File: tests/test_profiles/test_model/test_who_has_perm.py [hunk 18]
```diff
@@ -0,0 +1,208 @@
...
+        self.assertNotIn(self.user, user_with_permissions)
+        self.assertNotIn(self.user_with_no_perm, user_with_permissions)
+
+        ## Add view_tower_server_role in globalperm for self.user
+        self.global_perm.add_user_in_role(self.user, self.view_tower_server_role)
+
+        ## user is now in the list
+        user_with_permissions = self.tower_server.who_has_perm("service_catalog.view_towerserver")
+        self.assertIn(self.user, user_with_permissions)
+        self.assertNotIn(self.user_with_no_perm, user_with_permissions)
```
Comment (EliasBoulharts) [inline tests/test_profiles/test_model/test_who_has_perm.py:208]: Split into two files for a clearer overview

File: profiles/models/request_notification.py (hunk 1 above)
Comment (a-belhadj) [reply profiles/models/request_notification.py:46]: We have two types of notifications, one for instance that contains the squest instance in the context and one for request (this one) that contains only the squest request in his context. The goal here is to be able to use an InstanceNotificationFilter in a RequestNotificationFilter. If you don't agree, remove it :)

Comment (a-belhadj) [reply profiles/models/request_notification.py:46]: We have two types of notifications, one for instance that contains the squest instance in the context and one for request (this one) that contains only the squest request in his context. The goal here is to be able to use an InstanceNotificationFilter in a RequestNotificationFilter. If you don't agree, remove it :)

Comment (a-belhadj) [reply profiles/models/request_notification.py:46]: We have two types of notifications, one for instance that contains the squest instance in the context and one for request (this one) that contains only the squest request in his context. The goal here is to be able to use an InstanceNotificationFilter in a RequestNotificationFilter. If you don't agree, remove it :)

File: service_catalog/api/views/approval_workflow_api_views.py (hunk 2 above)
Comment (a-belhadj) [reply service_catalog/api/views/approval_workflow_api_views.py:53]: Ok for that

Comment (a-belhadj) [reply service_catalog/api/views/approval_workflow_api_views.py:53]: Parenthesis are missing here

File: service_catalog/api/views/approval_workflow_api_views.py (hunk 3 above)
Comment (a-belhadj) [reply service_catalog/api/views/approval_workflow_api_views.py:66]: Change it to a property please

File: service_catalog/api/views/approval_workflow_api_views.py (hunk 5 above)
Comment (a-belhadj) [reply service_catalog/api/views/approval_workflow_api_views.py:88]: It's an error, it should be

File: service_catalog/api/views/request_state_machine_api_view.py [hunk 19]
```diff
@@ -1,4 +1,5 @@
-from django.http import QueryDict
+import copy
```
[the remark this answers is not in the document]
Comment (a-belhadj) [reply service_catalog/api/views/request_state_machine_api_view.py:1]: I agree, please do it

File: tests/test_service_catalog/test_api/test_approval/test_approval_workflow_api_views.py (hunk 6 above)
Comment (a-belhadj) [reply tests/test_service_catalog/test_api/test_approval/test_approval_workflow_api_views.py:104]: Ok for that

File: tests/test_service_catalog/test_api/test_approval/test_approval_workflow_api_views.py (hunk 7 above)
Comment (a-belhadj) [reply tests/test_service_catalog/test_api/test_approval/test_approval_workflow_api_views.py:111]: Ok for that

File: tests/test_service_catalog/test_api/test_approval/test_approval_workflow_api_views.py [hunk 20]
```diff
@@ -66,3 +69,104 @@ def test_set_position_missing_position(self):
...
+        self.client.force_login(self.superuser)
+        url_approve = reverse('api_request_approval_workflow_state_approve', kwargs={'pk': self.test_request.pk})
+        response = self.client.post(url_approve, format='json')
+        self.test_request.refresh_from_db()
+        self.assertTrue(self.test_request.state == RequestState.ACCEPTED)
+        self.assertEqual(response.status_code, 200)
+
+    def test_superuser_can_reject(self):
+
+
```
[the remark this answers is not in the document]
Comment (a-belhadj) [reply tests/test_service_catalog/test_api/test_approval/test_approval_workflow_api_views.py:121]: I forgot to format the document before my commit

File: tests/test_mail_utils.py (hunk 13 above)
Comment (a-belhadj) [reply tests/test_mail_utils.py:215]: It can be split I agree

File: service_catalog/mail_utils.py (hunk 15 above)
Comment (a-belhadj) [reply service_catalog/mail_utils.py:32]: You're right, something like
```python
return qs.filter(email__isnull=False,profile__notification_filter_enabled=True) 
```

File: service_catalog/mail_utils.py (hunk 17 above)
Comment (a-belhadj) [reply service_catalog/mail_utils.py:34]: We can do it with a filter and then a values_list. Do it please

Comment (EliasBoulharts) [reply service_catalog/mail_utils.py:34]: We can not because of is_notifications_authorized method is called.

**Pull request merged**